### PR TITLE
Introduce `cast_starboard_api`.

### PR DESCRIPTION
**Pull request by saqneo**
The Cast Starboard API is a shared library which contains the portion of Starboard required to run Cast.

This includes the following changes:
- Add reference implementation, tests, and README.md into `starboard/contrib/cast` 
- Add `cast` to `//starboard:gn_all`
- Fix a bug in ApplicationX11 which was preventing it from working in modular builds, depending on invocation

Bug: b/281681333
Test: Reference implementation used on linux built of Cast (manual)

### CODE REVIEW

File: starboard/BUILD.gn [hunk 1]
```diff
@@ -177,24 +178,8 @@ source_set("starboard_headers_only") {
...
-    if (sb_is_evergreen_compatible) {
-      public_deps += [ "//third_party/crashpad/wrapper" ]
-    } else {
-      public_deps += [ "//third_party/crashpad/wrapper:wrapper_stub" ]
-    }
-    if (!sb_is_modular) {
-      public_deps += [ "//$starboard_path:starboard_platform" ]
-    }
+  # This is the default 'starboard_platform_target'; use default properties
+  starboard_platform_target("starboard_platform_group") {
```
Comment (kaidokert) [inline starboard/BUILD.gn:182]: @andrewsavage1 can you have a look ? This seems like an okay refactor here.

File: starboard/shared/x11/application_x11.cc [hunk 2]
```diff
@@ -919,7 +919,12 @@ ApplicationX11::WaitForSystemEventWithTimeout(SbTime time) {
     return pending_event;
   }
 
-  SB_DCHECK(dev_input_);
+  // In modular builds, |CreateWindow| is not always called before the event
+  // loop is running.
+  if (!dev_input_) {
```
Comment (kaidokert) [inline starboard/shared/x11/application_x11.cc:924]: The code seems fine here, but i wonder if we have other related surprises lurking somewhere. Did we miss some synchronization mechanism that should be happening ? No action needed here though, but maybe we need a quick follow-up bug to investigate.

File: starboard/BUILD.gn [hunk 3]
```diff
@@ -24,6 +24,7 @@ group("gn_all") {
     "//starboard/client_porting/eztime:eztime_test",
     "//starboard/client_porting/icu_init",
     "//starboard/client_porting/poem:poem_unittests",
+    "//starboard/contrib/cast/cast_starboard_api/samples:cast",
```
Comment (kaidokert) [inline starboard/BUILD.gn:27]: I don't know if we want to unconditionally include anything in `/contrib/` by default in the `all` build profile. I think this should perhaps be gated on a GN arg e.g. `contrib_build_with_cast` or such ? @y4vor @andrewsavage1 thoughts ?

File: starboard/contrib/cast/cast_starboard_api/samples/BUILD.gn [hunk 4]
```diff
@@ -0,0 +1,51 @@
+# Copyright 2023 The Cobalt Authors. All rights reserved.
+import("//starboard/build/config/starboard_target_type.gni")
+
+group("cast") {
+  if (build_with_separate_cobalt_toolchain) {
```
Comment (andrewsavage1) [inline starboard/contrib/cast/cast_starboard_api/samples/BUILD.gn:5]: Keep an eye on #1111  and #1181, maybe #1169 as well

File: starboard/build/config/starboard_target_type.gni [hunk 5]
```diff
@@ -23,3 +23,30 @@ if (starboard_target_type == "") {
     starboard_target_type = "group"
   }
 }
+
+template("starboard_platform_target") {
+  target(starboard_target_type, target_name) {
+    forward_variables_from(invoker, [ "ldflags" ])
```
Comment (andrewsavage1) [inline starboard/build/config/starboard_target_type.gni:29]: Instead of forwarding ldflags, put the cast ldflags in a config and add it to a list `extra_configs` which we forward here, then add to existing configs. See examples here: https://source.chromium.org/search?q=case:y%20%5CWextra_configs%5CW&sq=&ss=chromium%2Fchromium%2Fsrc

File: starboard/BUILD.gn (hunk 3 above)
Comment (andrewsavage1) [reply starboard/BUILD.gn:27]: Depends what cast wants to build on—if they only care about linux modular builds, that's fine, but if they care about many other platforms not building by default could cause our build matrix to blow up

File: starboard/contrib/cast/cast_starboard_api/samples/BUILD.gn [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+    public_deps = [ ":cast_starboard_api($starboard_toolchain)" ]
+  }
+}
+
+if (build_with_separate_cobalt_toolchain) {
+  if (current_toolchain == starboard_toolchain) {
+    starboard_platform_target("cast_starboard_api") {
+      ldflags = [
+        # Hide unwanted symbols, which also shrinks the resulting binary.
+        "-Wl,--version-script=" +
```
Comment (andrewsavage1) [inline starboard/contrib/cast/cast_starboard_api/samples/BUILD.gn:15]: Are these flags that could potentially go into the existing starboard_platform_group?

**Review by andrewsavage1 — CHANGES_REQUESTED** [review]

File: starboard/BUILD.gn (hunk 3 above)
Comment (saqneo) [reply starboard/BUILD.gn:27]: We are using the modular build by design as libcast_starboard_api.so will always be linked dynamically and is a thin wrapper on top of libstarboard_platform_group.so.

I'll make it so this is only included in the modular build.

File: starboard/contrib/cast/cast_starboard_api/samples/BUILD.gn (hunk 6 above)
Comment (saqneo) [reply starboard/contrib/cast/cast_starboard_api/samples/BUILD.gn:15]: > Are these flags that could potentially go into the existing starboard_platform_group?

I'm not sure how this would affect the default starboard_platform_group; limiting the externs like this actually reduces binary size by around ~10MB, so I don't feel comfortable doing it to the default lib.

Comment (andrewsavage1) [reply starboard/contrib/cast/cast_starboard_api/samples/BUILD.gn:15]: I believe we already do something very similar for Evergreen, @y4vor could comment

File: starboard/contrib/cast/cast_starboard_api/samples/BUILD.gn (hunk 4 above)
Comment (saqneo) [reply starboard/contrib/cast/cast_starboard_api/samples/BUILD.gn:5]: Thanks for the heads up; I believe I have these integrated already and they are not affecting the binary at the moment.

**Comment by andrewsavage1** [comment]
I think this is all good, but get @y4vor to look at the comment I added about potentially using your starboard flags ourselves (it's fine to submit this then follow up with that later). Also let's test this internally in https://lbshell-internal-review.git.corp.google.com/c/cobalt_src/+/260424, let me know if you need help with that

File: starboard/BUILD.gn [hunk 7]
```diff
@@ -57,9 +57,13 @@ group("gn_all") {
...
-  if (build_with_separate_cobalt_toolchain && is_cobalt_toolchain) {
-    assert(sb_is_modular,
-           "sb_is_modular should be set when building with cobalt_toolchain")
+  if (build_with_separate_cobalt_toolchain) {
+    if (is_cobalt_toolchain) {
+      assert(sb_is_modular,
+             "sb_is_modular should be set when building with cobalt_toolchain")
+    }
+
+    deps += [ "//starboard/contrib/cast/cast_starboard_api/samples:cast" ]
```
Comment (y4vor) [inline starboard/BUILD.gn:66]: You would need a new variable. We don't want all modular builds to have cast in them.
Something like `cast_sample`?

File: starboard/BUILD.gn [hunk 8]
```diff
@@ -177,24 +181,8 @@ source_set("starboard_headers_only") {
 }
 
 if (current_toolchain == starboard_toolchain) {
-  target(starboard_target_type, "starboard_platform_group") {
```
Comment (y4vor) [inline starboard/BUILD.gn:180]: Was there anything functionally wrong with the template?
If not, let's not change it here. Even if the refactoring is a noop good cleanup it should be in separate PR.

File: starboard/contrib/cast/README.md [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+return in the future.
+
+Cast still requires additional behavior be implemented behind the existing
+Starboard APIs. Reference the `Cast TV Integration Guide` for details.
+
+### Reference Implementation
+
+The `cast_starboard_api/samples/` directory contains the reference target
+`cast_starboard_api` which can be built when
+`build_with_separate_cobalt_toolchain=true` is specified. To generate the
```
Comment (y4vor) [inline starboard/contrib/cast/README.md:20]: I would also add the new variable. So something like
`build_with_separate_cobalt_toolchain=true` `cast_sample=true`

File: starboard/shared/x11/application_x11.cc (hunk 2 above)
Comment (y4vor) [reply starboard/shared/x11/application_x11.cc:924]: +1 to file a ticket with repro steps.

File: starboard/BUILD.gn (hunk 8 above)
Comment (andrewsavage1) [reply starboard/BUILD.gn:180]: Extra configs need to be passed to this target for the build to work, it needs to be done in this PR

File: starboard/BUILD.gn (hunk 7 above)
Comment (andrewsavage1) [reply starboard/BUILD.gn:66]: GN variables should follow https://gn.googlesource.com/gn/+/main/docs/style_guide.md#Naming-conventions. If we introduce a variable for this, it should likely be prefixed with `enable_` or `use_`

File: starboard/contrib/cast/README.md (hunk 9 above)
Comment (saqneo) [reply starboard/contrib/cast/README.md]: Added `use_contrib_cast`

File: starboard/BUILD.gn (hunk 7 above)
Comment (saqneo) [reply starboard/BUILD.gn]: Added `use_contrib_cast`

File: starboard/shared/x11/application_x11.cc (hunk 2 above)
Comment (saqneo) [reply starboard/shared/x11/application_x11.cc:924]: Opened an internal bug with repro steps.

File: starboard/contrib/cast/cast_starboard_api/samples/BUILD.gn (hunk 6 above)
Comment (kaidokert) [reply starboard/contrib/cast/cast_starboard_api/samples/BUILD.gn]: I think @andrewsavage1 is out for next couple days - @niranjanyardi and @y4vor for a quick look. The rest of the PR seems to have everything addressed and ready to land.

**Review by y4vor — APPROVED** [review]

File: starboard/contrib/cast/cast_starboard_api/samples/BUILD.gn (hunk 6 above)
Comment (saqneo) [reply starboard/contrib/cast/cast_starboard_api/samples/BUILD.gn]: Resolving since @y4vor has approved. If these flags prove useful and stable for libstarboard_platform_group.so, they can be refactored in a separate change.
